dbms/include/DB/Storages/MergeTree/MergeTreeDataMerger.h

```diff
@@ -120,6 +121,19 @@ class MergeTreeDataMerger
 
 	bool isCancelled() const { return cancelled > 0; }
 
+public:
+
+	enum class MergeAlg
+	{
+		BASIC,
+		VERTICAL
+	};
+
+private:
+
+	MergeAlg chooseMergingAlg(const MergeTreeData & data, const MergeTreeData::DataPartsVector & parts,
+		size_t rows_upper_bound, MergedRowSources & rows_sources_to_alloc) const;
```

Comment:
It was motivated by the following logic: If we cannot reserve memory -> choose basic alg